pkg/sdn/plugin/egress_dns.go

```diff
@@ -0,0 +1,166 @@
+package plugin
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	osapi "github.com/openshift/origin/pkg/sdn/api"
+
+	ktypes "k8s.io/kubernetes/pkg/types"
+)
+
+type EgressDNSUpdate struct {
+	UID       ktypes.UID
+	Namespace string
+}
+
+type EgressDNS struct {
+	// Protects pdMap/namespaces operations
+	lock sync.Mutex
+	// Holds Egress DNS entries for each policy
+	pdMap map[ktypes.UID]*DNS
+	// Maintain namespaces for each policy to avoid querying etcd in syncEgressDNSPolicyRules()
+	namespaces map[ktypes.UID]string
+
+	// Report change when Add operation is done
+	added chan bool
+
+	// Report changes when there are dns updates
+	updates chan EgressDNSUpdate
+}
+
+func (e *EgressDNS) Add(policy osapi.EgressNetworkPolicy) {
+	dnsInfo := &DNS{}
+	for _, rule := range policy.Spec.Egress {
+		if len(rule.To.DNSName) > 0 {
+			if err := dnsInfo.Add(rule.To.DNSName); err != nil {
+				glog.Error(err)
+			}
+		}
+	}
+
+	if dnsInfo.Size() > 0 {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		if len(e.pdMap) == 0 {
+			e.pdMap = map[ktypes.UID]*DNS{}
+			e.namespaces = map[ktypes.UID]string{}
+		}
+		e.pdMap[policy.UID] = dnsInfo
+		e.namespaces[policy.UID] = policy.Namespace
+		e.signalAdded()
+	}
+}
+
+func (e *EgressDNS) Delete(policy osapi.EgressNetworkPolicy) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if _, ok := e.pdMap[policy.UID]; ok {
+		delete(e.pdMap, policy.UID)
+		delete(e.namespaces, policy.UID)
+	}
+}
+
+func (e *EgressDNS) Update(policyUID ktypes.UID) (error, bool) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if dnsInfo, ok := e.pdMap[policyUID]; ok {
+		return dnsInfo.Update()
+	}
+	return nil, false
+}
+
+func (e *EgressDNS) SyncInit() {
```

Comment:
It would be more idiomatic to have a NewEgressDNS() method that returns a fully-initialized object, rather than using &EgressDNS{} from node/proxy and then having to call SyncInit()